Use molar density in material description

msmk0 opened the pull request:
Use molar material density (number-of-atoms/volume e.g. in mol/cm³) instead of mass density (mass/volume e.g. in g/cm³) for the material description. This avoids numerical issues due to our chosen native mass unit (GeV) and is better suited for the computations that are actually performed.

Adds `mol` to the unit definitions.

Fixes #31.

msmk0 commented:
@Corentin-Allaire Feel free to have a look. However, I think that this needs a thorough review/discussion that can not be done in an afternoon. This adds quite a few changes that are probably better handled in separate PRs. One particular issue is that when persisting the material properties (Json material map, ROOT material track writer), it should be done using the encoded parameters vector without caring about its content.

For now I will leave this *WIP* to clarify that this needs further discussion.

Corentin-Allaire commented:
Ok, I will try to have a quick look but seeing the size of this PR we should probably discuss it in more detail in two week.

msmk0 commented:
@Corentin-Allaire I rebase this PR and fixed the remaining build and test issues. From my side it is ready to be reviewed in detail.

The major point that will probably need discussion are the changes to the JSON converters. They were using the various property accessors explicitely and I switched them to use only the opaque classification numbers. For storing the material properties it should not matter which element in the parameters vector is which and what its exact definition is. In all our code base these values are only used through the `Material` class anyways (and should). This change can break some of the existing material maps and could be split in a separate PR.

Corentin-Allaire commented:
@msmk0 I am in vacation in Brittany until Friday next week but can try to review this early next week.

msmk0 commented:
@Corentin-Allaire My apologies, I was not aware. This is not time critical and can wait until you are back. I will switch it back to WIP until then.

msmk0 commented:
@Corentin-Allaire I think we should try to get this into `v1.0.0` (or decide against it altogether). I rebased it onto the latest master. I will leave this as WIP until we have tagged `v0.32.0` to avoid confusion, but you can already have a look.

Corentin-Allaire commented:
@msmk0 It was my plan to look at it today. But I agree we should had this to 1.0.0

Corentin-Allaire commented on Core/src/Material/MaterialProperties.cpp:
```diff
@@ -17,20 +17,12 @@ static constexpr auto eps = 2 * std::numeric_limits<float>::epsilon();
-      m_thicknessInX0((material.X0() > eps) ? (thickness / material.X0()) : 0),
-      m_thicknessInL0((material.L0() > eps) ? (thickness / material.L0()) : 0) {
+      m_thicknessInX0((eps < material.X0()) ? (thickness / material.X0()) : 0),
+      m_thicknessInL0((eps < material.L0()) ? (thickness / material.L0()) : 0) {
```
Really minor point but what is the reasoning for changing the ordering here ? In my head the norm for writing comparison is "variable <, > or == constant".

msmk0 replied:
I always try to use `<`-only for order comparison: this orders the components the same way you would order them along a number axis. I find this particularly convenient e.g. for limit checks on both side
```
if ((0 < x) and (x < 1)) { ... }
```
With this notation it is also visually clear that `x` is within the lower and upper limits.

Of course, this is a personal preference.

Corentin-Allaire replied:
I was mostly just curious, that is an interesting way to look at this. Thanks for the answer !

Corentin-Allaire reviewed (commented):
Great PR, correct me if I am wrong but rho isn't actually used anywhere right now ? So it shouldn't caused too much issue (I will just need to redo material map afterward).

BTW I love that you actually wrote the math from mass-density to molar density in comments. We should advertise doing the same whenever it is possibe in our contributing guide I feel !

msmk0 commented on Core/include/Acts/Material/Material.hpp:
```diff
@@ -24,25 +24,47 @@ namespace Acts {
+  ///   can lead to numerical problems. Typical mass densities lead to
+  ///   computations with values differing by 20+ orders of magnitude.
+  static Material fromMassDensity(float x0, float l0, float ar, float z,
+                                  float rho);
```
Good point. I renamed this parameter to `massRho` for consistency.

msmk0 commented:
> Great PR, correct me if I am wrong but rho isn't actually used anywhere right now ? So it shouldn't caused too much issue (I will just need to redo material map afterward).

Mass density in the core is still used for averaging materials in a few cases. I want to replace those before this gets merged. That will probably also solve #30 and thus obsolete #292 as well.

The `.massDensity()` accessor should remain available. That quantity is used in many other packages (e.g. Geant4) and it makes sense to have easy access to it.

Corentin-Allaire commented:
> > Great PR, correct me if I am wrong but rho isn't actually used anywhere right now ? So it shouldn't caused too much issue (I will just need to redo material map afterward).
> 
> Mass density in the core is still used for averaging materials in a few cases. I want to replace those before this gets merged. That will probably also solve #30 and thus obsolete #292 as well.
> 
> The `.massDensity()` accessor should remain available. That quantity is used in many other packages (e.g. Geant4) and it makes sense to have easy access to it.

A yes you are right it is still used by the `AccumulatedMaterialProperties` to average Ar and Z and should indeed be replace by the molarDensity there. But since neither Ar, Z or the massDensity are explicitly used anywhere this change shouldn't affect any result (just the value writen by the mapping).

Corentin-Allaire commented:
I now look great and clearer than ever ! The only thing I still have some doubt about is the use of the path correction. I think you do it correctly but if you don't mind I would like to take this branch and use it to map the ODD, that way we can be sure that everything is still fine. I will do that on Monday and approve after.

msmk0 commented:
Yes, that is a good idea.

Corentin-Allaire commented:
Great news, I mapped the ODD with this branch and the result is identical to what we have in master : 
master :
[Ratio_Val_geant_mat_X0_Eta.pdf](https://github.com/acts-project/acts/files/5150446/Ratio_Val_geant_mat_X0_Eta.pdf)

PR :
[Ratio_Val_geant_mat_X0_Eta.pdf](https://github.com/acts-project/acts/files/5150452/Ratio_Val_geant_mat_X0_Eta.pdf)

I does create some pb with the json map it self if we have vacuum bin, we used to write no material :
[
]
But now we write vacuum : 
[
null,
null,
0,
0,
0,
0
]
Which cannot be read back by json. I will see what we could change either to the converter or the accumulator.

msmk0 commented:
The first two entries contain X0/L0 and should both be `inf` for vacuum., but JSON does not support encoding `inf`. I would guess that the code does the right thing here but the JSON encoding does not support vacuum entries.

Corentin-Allaire commented:
> The first two entries contain X0/L0 and should both be `inf` for vacuum., but JSON does not support encoding `inf`. I would guess that the code does the right thing here but the JSON encoding does not support vacuum entries.
> 
> Could we encode vacuum directly as `null` instead of encoding the parameters explicitely, i.e. only write out the parameters if we actually have real matter?

That is what we use to do actually

Corentin-Allaire commented on Plugins/Json/src/JsonGeometryConverter.cpp:
```diff
@@ -611,17 +611,10 @@ json Acts::JsonGeometryConverter::surfaceMaterialToJson(
-    }
-    return {};
+    const auto cn = mp.material().parameters();
+    return {cn[0], cn[1], cn[2], cn[3], cn[4], mp.thickness()};
```
Here if(mp.material() == Acts::Material()) you should return {}. This would solve our issue.

msmk0 replied:
What do we do for finite vacuum? `[nullptr, thickness]`?

Corentin-Allaire replied:
Yes, or we just write the thickness and add a if (size == 1) to then jsonToMaterialMatrix.

msmk0 replied:
I added a proposal for updating the encoding/decoding to avoid any ambiguity.

Indefinite material is encoded either as a JSON array `[1, 2, 3, 4, 5]` or directly as `null` in case of vacuum. A concrete material slab is always encoded as a JSON object with two members
```
{
  "material": [1, 2, 3, 4, 5], // or null in case of vacuum
  "thickness": 1.2,
}
```
This allows encoding of all special cases (vacuum w/ zero thickness, vacuum w/ non-zero thickness, real material w/ zero thickness) without loss of information at the expense of a slightly more verbose output.

This way the `Material` encoding is the same in both cases and indefinite and concrete materials are clearly encoded in two different ways.

Corentin-Allaire commented on Plugins/Json/src/JsonGeometryConverter.cpp:
```diff
@@ -720,6 +714,11 @@ json Acts::JsonGeometryConverter::surfaceMaterialToJson(
     const Acts::IVolumeMaterial& vMaterial) {
+  auto encodeMaterialParameters =
+      [](const Material::ParametersVector& params) -> std::vector<float> {
+    return {params[0], params[1], params[2], params[3], params[3]};
```
Same here.

msmk0 replied:
No problem here. I will change this. There is also a bug in that line: `param[4]` is never stored.

Corentin-Allaire replied:
Technically you don't need `param[4]`, IVolumeMaterial use material and not MaterialProperties so there is no thickness.

msmk0 replied:
`param[4]` is the molar density and still part of the `Material` description.

Corentin-Allaire replied:
Ah yes you are right !

msmk0 replied:
See above. I will close this conversation for now.

Corentin-Allaire commented on Plugins/Json/src/JsonGeometryConverter.cpp:
```diff
@@ -394,8 +447,7 @@ const Acts::IVolumeMaterial* Acts::JsonGeometryConverter::jsonToVolumeMaterial(
   } else if (mmat.size() == 1) {
-    vMaterial = new Acts::HomogeneousVolumeMaterial(Acts::Material(
-        mmat[0][0], mmat[0][1], mmat[0][2], mmat[0][3], mmat[0][4]));
+    vMaterial = new Acts::HomogeneousVolumeMaterial(Acts::Material(mmat[0]));
```
Should use decodeMaterial() here.

msmk0 replied:
Missed that one, thanks.

Corentin-Allaire commented on Plugins/Json/src/JsonGeometryConverter.cpp:
```diff
@@ -412,9 +464,7 @@ const Acts::IVolumeMaterial* Acts::JsonGeometryConverter::jsonToVolumeMaterial(
-        mGrid.at(bin) = Acts::Material(mmat[bin][0], mmat[bin][1], mmat[bin][2],
-                                       mmat[bin][3], mmat[bin][4])
-                            .classificationNumbers();
+        mGrid.at(bin) = mmat[bin];
```
Same here.

msmk0 replied:
And that one.

Corentin-Allaire commented on Plugins/Json/src/JsonGeometryConverter.cpp:
```diff
@@ -436,9 +486,7 @@ const Acts::IVolumeMaterial* Acts::JsonGeometryConverter::jsonToVolumeMaterial(
-        mGrid.at(bin) = Acts::Material(mmat[bin][0], mmat[bin][1], mmat[bin][2],
-                                       mmat[bin][3], mmat[bin][4])
-                            .classificationNumbers();
+        mGrid.at(bin) = mmat[bin];
```
And also here.

msmk0 replied:
And that one too :wink:

Corentin-Allaire reviewed (commented):
I like your new encoding for the material / thickness.
The only thing is that if you write volume material it does not work. Changing the line I mentioned below should fix it.

Is there anything else you would like to change for this PR ? If not I will go through it one last then approve it.

msmk0 commented:
> Is there anything else you would like to change for this PR ? If not I will go through it one last then approve it.

No, I am just fixing the failure and then it can go in. Do you want to quickly test the material mapping again after I added the fixes?

Corentin-Allaire commented:
Yes, it will mapped the material again just to be sure. I will also test the volume mapping this time, I am mapping the long strip of the ODD as a volume to check that this also work.

msmk0 commented:
@Corentin-Allaire I fixed the outstanding issue and updated the material map JSON converter test. The test now uses input data from an example file in the `Tests/Data` directory.

Corentin-Allaire commented:
I am a bit confused, I get an error in the propagation when I have volume material, any idea of why that could be ?

```
Info in <TGeoManager::CloseGeometry>: ----------------modeler ready----------------
Detector         INFO  +++ Patching names of anonymous shapes....
DD4hepVolumeManager INFO  **** running plugin DD4hepVolumeManager ! 
VolumeManager    INFO   - populating volume ids - be patient ...
VolumeManager    INFO   - populating volume ids - done. 19808 nodes.
VolumeManager    INFO  +++ Volume manager populated and loaded.
17:08:34    DD4hepConver   INFO      Translating DD4hep geometry into Acts geometry
17:08:34    DD4hepConver   INFO      Translating DD4hep sub detector: BeamPipe
17:08:34    DD4hepConver   INFO      Translating DD4hep sub detector: Pixels
17:08:34    DD4hepConver   INFO      Translating DD4hep sub detector: ShortStrips
17:08:34    DD4hepConver   INFO      Translating DD4hep sub detector: LongStrips
- BField (scalor to/in) Tesla set to: 1
17:08:34    Sequencer      INFO      Added algorithm 'PropagationAlgorithm'
17:08:34    Sequencer      INFO      Added writer 'RootMaterialTrackWriter'
17:08:34    Sequencer      INFO      Processing events [0, 1000)
17:08:34    Sequencer      INFO      Starting event loop with 1 threads
17:08:34    Sequencer      INFO        0 services
17:08:34    Sequencer      INFO        0 context decorators
17:08:34    Sequencer      INFO        0 readers
17:08:34    Sequencer      INFO        1 algorithms
17:08:34    Sequencer      INFO        1 writers
17:08:34    PropagationA   ERROR     Propagation reached the step count limit of 1000(did 1000 steps)
TBB Warning: Exact exception propagation is requested by application but the linked library is built without support for it

**************************************************** 
*  A runtime error has occured :                     
*    Value called on error value: PropagatorError: unknown [3]
*  the program will have to be terminated - sorry.   
**************************************************** 
```

paulgessinger commented:
Navigation failure?

msmk0 commented:
I think propagator error 3 indicates too many steps. I had to increase the maximum number of steps for some of the integration tests as well. Not sure why.

Corentin-Allaire commented:
After discussing with Fabian increasing the amount of material could increase the number of step (but the material should be the same). I will try to do some test to see if the material seen during the propagation make sense.

Corentin-Allaire commented:
After looking at it in a bit more detail it seems to only occur in the end cap which is why it was not happening to me before. The step size seem to go as low as 10e-6 so there might be an issue somewhere in the step size computation...
But this PR should be fine, let me have a final look then I can approve it.

msmk0 commented:
@FabianKlimpel @Corentin-Allaire This happens in the dense propagator, right? How exactly does this use the material to change the step size? Does it use e.g. the mass density directly or only indirectly via the resulting energy loss?

FabianKlimpel commented:
> @FabianKlimpel @Corentin-Allaire This happens in the dense propagator, right? How exactly does this use the material to change the step size? Does it use e.g. the mass density directly or only indirectly via the resulting energy loss?

The step size is adjusted based on the integration error (https://github.com/acts-project/acts/blob/master/Core/include/Acts/Propagator/EigenStepper.ipp#L156). So it is an indirect effect based on the q/p change along the step.

paulgessinger commented:
In principle, the maximum number of steps is there as a safeguard against an accidental infinite loop. If you can get it to not error-out by increasing the step limit, that should be ok I think.
